File: 2015/03/solution.py
def next_house(house: tuple[int], move: str) -> tuple:
    x, y = house
    match move:
        case '^':
            y += 1
        case 'v':
            y -= 1
        case '>':
            x += 1
        case '<':
            x -= 1
    return x, y


line = input()
robo_house = santa2_house = santa1_house = 0, 0
year1 = {santa1_house}
year2 = year1.copy()
for i, move in enumerate(line):
    santa1_house = next_house(santa1_house, move)
    year1.add(santa1_house)
    if i % 2 == 0:
        santa2_house = next_house(santa2_house, move)
        year2.add(santa2_house)
    else:
        robo_house = next_house(robo_house, move)
        year2.add(robo_house)
print("Part One:", len(year1))
print("Part Two:", len(year2))

File: 2015/03/solution.go
package main

import (
	"fmt"
)

func nextPos(pos []int, move rune) {
	switch move {
	case '^':
		pos[1]++
	case 'v':
		pos[1]--
	case '>':
		pos[0]++
	case '<':
		pos[0]--
	}
}

func main() {
	var line string
	var santa1Pos, santa2Pos, roboSantaPos [2]int
	year1 := make(map[[2]int]bool)
	year2 := make(map[[2]int]bool)
	fmt.Scanln(&line)
	year1[santa1Pos] = true
	year2[santa2Pos] = true
	for i, move := range line {
		nextPos(santa1Pos[:], move)
		year1[santa1Pos] = true
		if i%2 == 0 {
			nextPos(santa2Pos[:], move)
			year2[santa2Pos] = true
		} else {
			nextPos(roboSantaPos[:], move)
			year2[roboSantaPos] = true
		}
	}
	fmt.Println("Part One:", len(year1))
	fmt.Println("Part Two:", len(year2))
}
